docs: storage and state variables (#1725)
AztecProtocol/aztec-packages: docs/docs/dev_docs/contracts/state_variables.md

# State Variables

State variables come in two flavours: [**public** state](#publicstatet-t_serialised_len) and [**private** state](#private-state-variables).

> 

Reviewer: (This can be a separate PR, though :) )

> 

Author: #1856

### `::new`

Here we define a singleton for storing a Card note:

#include_code state_vars-Singleton /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/storage/legendary_card.nr rust

### `.initialise`

The initial value of a singleton is set via calling `initialise`:

#include_code state_vars-SingletonInit /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/actions.nr rust

When this function is called, a nullifier of the storage slot is created, preventing this singleton from being initialised again.

### `.replace`

To modify the value of a singleton, we will create a note (a Card in the following example) that has the new data, and replace the current note with it:

#include_code state_vars-SingletonReplace /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/actions.nr rust

This function will destroy the old note under the hood. If two people are trying to modify the singleton at the same time, only one will succeed.

### `.get_note`

This function allows us to get the note of a singleton:

#include_code state_vars-SingletonGet /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/actions.nr rust

> However, it's possible that at the time this function is called, the system hasn't synched to the block where the latest note was created. Or a malicious user might feed an old state to this function, tricking the proving system into thinking that the value hasn't changed. To avoid an attack around it, this function will destroy the current note, and replace it with a duplicated note that has the same fields. Because the nullifier of the latest note will be emitted, if two people are trying to use this function against the same note, only one will succeed.

Author: `get_note` will destroy the current note and create a new note, and `replace` will destroy the new note and create the replacement note. So it should work just fine. Although the extra read request and nullifier are really annoying.
But I don't know if it's worth teaching people the differences of the apis to save 1 extra slot in MAX___PER_CALL. 

What if we need to call `replace` twice in a call:
```rust
let mut note = singleton.get_note_as_mutable();
for i in 0..4 {
  let score = contractB.play_round()
  if score > some_value {
    note.count += 1;
    var.replace(note);
  }
}
```
> I guess we can reset `_already_got` (maybe rename it to `_nullified`?) after the replacement note is created!

Or, what if a nested function also needs to access the singleton:
```rust
fn A() {
  let mut note = singleton.get_note_as_mutable();
  if note.owner == some_value {
    this.B()
  }
  note.count += 1;
  singleton.replace(note);
}

fn B() {
  let note = singleton.get_note_as_read_only();
  if note.owner == some_value {
    do_something();
  }
}
```
Function A will nullify the current note. And function B will also trying to nullify the current note (same nullifier!), and create a duplicated note. Then function A will create the replacement note.

> However, it's possible that at the time this function is called, the system hasn't synched to the block where the latest note was created. Or a malicious user might feed an old state to this function, tricking the proving system into thinking that the value hasn't changed. To avoid an attack around it, this function will destroy the current note, and replace it with a duplicated note that has the same fields. Because the nullifier of the latest note will be emitted, if two people are trying to use this function against the same note, only one will succeed.

Reviewer: Sorry, I'm a bit delayed catching up.

> What if we need to call replace twice in a call:

That example is probably just something we should teach as a bad pattern. The same can be achieved more efficiently if `.replace` is called only at the end, instead of every iteration.

> Or, what if a nested function also needs to access the singleton:

In that example, the function `B` could be revised to take the note as an argument.

HERE'S A THOUGHT...

To help users, the `context` could track information about each read, nullification, and write, and only emit read_requests / nullifiers / new commitments at the time that `context.finalise()` is called (or at the time an external `call` is made). That would be cool.

> However, it's possible that at the time this function is called, the system hasn't synched to the block where the latest note was created. Or a malicious user might feed an old state to this function, tricking the proving system into thinking that the value hasn't changed. To avoid an attack around it, this function will destroy the current note, and replace it with a duplicated note that has the same fields. Because the nullifier of the latest note will be emitted, if two people are trying to use this function against the same note, only one will succeed.

Author: #1898

### `.get_notes`

This function returns the notes the account has access to:

#include_code state_vars-SetGet /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/actions.nr rust

There's a limit on the maxinum number of notes this function can return at a time. Check [here](INSERT_LINK_HERE) and look for `MAX_READ_REQUESTS_PER_CALL` for the up-to-date number.

Because of this limit, we should always consider using the second argument `NoteGetterOptions` to target the notes we need, and to reduce the time required to recursively call this function.

### NoteGetterOptions

`NoteGetterOptions` encapsulates a set of configurable options for retrieving notes from a database:

#include_code NoteGetterOptions /yarn-project/noir-libs/noir-aztec/src/note/note_getter_options.nr rust

For example, the following configs let us find the cards that belong to `account_address`. The returned cards are sorted by their points in descending order, and the first `offset` cards with the highest points are skipped.

#include_code state_vars-NoteGetterOptionsSelectSortOffset /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/options.nr rust

> The first value of `.select` and `.sort` is the index of a field in a note type. For the note type `Card` that has the following fields:

Author: #1857

> #include_code state_vars-NoteCard /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/types/card.nr rust

Author: That's what I did. But then everything is a note. It got a little bit overwhelming in the end 😂

#include_code state_vars-OptionFilter /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/options.nr rust

We can use it as a filter to further reduce the number of the final notes:

#include_code state_vars-NoteGetterOptionsFilter /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/options.nr rust

One thing to remember is, `filter` will be applied on the notes after they are picked from the database. Therefor, it's possible that the actual notes we end up getting are fewer than the limit.

The limit is `MAX_READ_REQUESTS_PER_CALL` by default. But we can set it to any value "smaller" than that:

#include_code state_vars-NoteGetterOptionsPickOne /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/options.nr rust

The process of applying the options to get the final notes is not constrained. It's necessary to always check the returned notes even when some conditions have been specified in the options.

#include_code state_vars-check_return_notes /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/main.nr rust

## `Map<T>`

> `Map` is a state variable that maps a field to another state variable, which can be [`PublicState`](#publicstatet-t_serialised_len), all the [private state variables](#private-state-variables), and even another Map.

Reviewer: ```suggestion
`Map` is a state variable that maps a `Field` to another state variable, which can be [`PublicState`](#publicstatet-t_serialised_len), all the [private state variables](#private-state-variables), and even another Map.

> `Map` can map from `Field` or any native Noir type which is convertible to `Field`.
```

### `::new`

> The following map uses singleton as its value:

Reviewer: ```suggestion
The following declares a mapping from a `Field` to a `Singleton`:
```

#include_code state_vars-MapSingleton /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/storage/profiles.nr rust

> 

Reviewer: ```suggestion
The second argument `|slot| Singleton::new(slot, ProfileMethods)` is a Noir closure function. It teaches this instance of `Map` how to create a new instance of a `Singleton` whenever the `.at` method is called to access a state variable at a particular mapping key. The `slot` argument will be derived when `.at` is called, based on the lookup key provided.
```
